metadata_annotator/history_functions.py

```
@@ -0,0 +1,197 @@
+import os
+import re
+from datetime import UTC, datetime
+from urllib.parse import parse_qs, unquote
+
+import xarray as xr
+
+PROGRAM = 'Harmony Metadata Annotator'
+# To be improved - make dynamic based on service_version.txt
+VERSION = '0.0.1'
+
+
+def update_history_metadata(datatree: xr.DataTree) -> None:
+    """Update the history global attribute of the DataTree.
+
+    If either an existing History or history global attribute exists, append
+    information as a new line to the existing value. Otherwise create a new
+    attribute called "history".
+    """
+    history_attribute_name, existing_history = read_history_attrs(datatree)
+    write_history_attrs(datatree, history_attribute_name, existing_history)
+
+
+def read_history_attrs(datatree: xr.DataTree) -> tuple[str, str]:
+    """Read history attribute."""
+    if 'History' in datatree.attrs:
+        history_attribute_name = 'History'
+        existing_history = datatree.attrs['History']
+    elif 'history' in datatree.attrs:
+        history_attribute_name = 'history'
+        existing_history = datatree.attrs['history']
+    else:
+        history_attribute_name = 'history'
+        existing_history = None
+    return history_attribute_name, existing_history
+
+
+def write_history_attrs(
+    datatree: xr.DataTree, history_attribute_name: str, existing_history: str
+) -> None:
+    """Write history attribute."""
+    new_history_line = ' '.join(
+        [
+            datetime.now(UTC).isoformat(),
+            PROGRAM,
+            VERSION,
+        ]
+    )
+    datatree.attrs[history_attribute_name] = '\n'.join(
+        filter(None, [existing_history, new_history_line])
+    )
+
+
+def get_start_index_from_history(
+    dimension_index_map: dict[str, int], dimension_variable_path: str
+) -> int:
+    """Return the start index from the dimension index map."""
+    if dimension_variable_path in dimension_index_map.keys():
+        return dimension_index_map[dimension_variable_path]
+    return 0
```

If a variable does not exist in dimension_index_map, should we raise an exception here instead of returning 0?

---

If dimension does not exist, it would be a case where there is no spatial subset and no index ranges are in the history attribute. The start index would be 0.